### PR TITLE
[Core][Model/ModelPart] recursive SubModelParts

### PR DESCRIPTION
**Pull request by philbucher**
:warning:  **PLEASE READ THIS BEFORE BLOCKING** :warning: 
I am really hesitant to change both `Model` and `ModelPart`, but after implementing those functionalities for the 20th time in my own scripts I felt like it is time to do it properly

In a nutshell with this PR it becomes possible to use inputs such as `Inlet.SubModelPartName` for `Create(Sub)ModelPart`, `Get(Sub)ModelPart` and `Has(Sub)ModelPart`

This is very handy when dealing with more complex `ModelPart` hierarchies as it is necessary for larger models

### CODE REVIEW

**Comment by philbucher** [comment]
(sorry for the whitespace changes)

**Review by loumalouomega — DISMISSED** [review]
For me this Ok as it improves the current behavior and does not break current one

File: kratos/sources/model_part.cpp [hunk 1]
```diff
@@ -1616,28 +1632,71 @@ void ModelPart::RemoveGeometryFromAllLevels(std::string GeometryName)
...
-        << "\" in model part: \"" << Name() << "\"" << std::endl;
+    const auto sub_model_parts_list = SplitSubModelPartHierarchy(NewSubModelPartName);
+    if (sub_model_parts_list.size() == 1) {
+        // Here a warning would be enough. To be disscussed. Pooyan.
+        KRATOS_ERROR_IF(mSubModelParts.find(NewSubModelPartName) != mSubModelParts.end())
+            << "There is an already existing sub model part with name \"" << NewSubModelPartName
+            << "\" in model part: \"" << Name() << "\"" << std::endl;
+
+        ModelPart* praw = new ModelPart(NewSubModelPartName, this->mpVariablesList, this->GetModel());
+        Kratos::shared_ptr<ModelPart> p_model_part(praw); //we need to construct first a raw pointer
```
Comment (adityaghantasala) [inline kratos/sources/model_part.cpp:1969]: Is this required ? first a raw and then a shared pointer ?

Comment (philbucher) [reply kratos/sources/model_part.cpp:1969]: pretty sure it is, @RiccardoRossi wrote this

Comment (philbucher) [reply kratos/sources/model_part.cpp:1969]: thats bcs `ModelPart` doesn't have a public constructor, and `unique_ptr` is not a `friend` of `ModelPart`, but `Model` is

File: kratos/tests/cpp_tests/containers/test_model.cpp [hunk 2]
```diff
@@ -18,125 +18,155 @@
...
+    // Checking SubSubModelPart
+    // here Inlet1 exists already
+    auto& ssmp = model.CreateModelPart("Main.Inlet1.sub_inlet");
+    KRATOS_CHECK(model_part.HasSubModelPart("Inlet1.sub_inlet"));
+    KRATOS_CHECK_EQUAL("sub_inlet", ssmp.Name());
 
+    // here InletCustom does NOT exists yet
+    auto& ssmp_2 = model.CreateModelPart("Main.InletCustom.ccc_sub_inlet");
+    KRATOS_CHECK(model_part.HasSubModelPart("InletCustom.ccc_sub_inlet"));
+    KRATOS_CHECK_EQUAL("ccc_sub_inlet", ssmp_2.Name());
```
Comment (armingeiser) [inline kratos/tests/cpp_tests/containers/test_model.cpp:161]: Can you also check if `Main.InletCustom` exists?

**Comment by armingeiser** [comment]
I think it is good to be this flexible, however the restriction that a model part name can not contain a `.` becomes more relevant now. It has to be documented properly that for those functions a `.` is used to split the string into a hierarchy of submodelpart names.

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee accepts this interface, however the code should be reviewed in detail before merging

File: kratos/containers/model.cpp [hunk 3]
```diff
@@ -33,27 +34,57 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
-    auto search = mRootModelPartMap.find(ModelPartName);
-    if( search == mRootModelPartMap.end()) {
-        auto pvar_list = Kratos::make_intrusive<VariablesList>();
+    KRATOS_ERROR_IF( ModelPartName.empty() ) << "Please don't use empty names (\"\") when creating a ModelPart" << std::endl;
 
-            ModelPart* pmodel_part = new ModelPart(ModelPartName, NewBufferSize, pvar_list, *this );
-        mRootModelPartMap[ModelPartName] = std::unique_ptr<ModelPart>(pmodel_part); //note that i create it separately since Model is friend of ModelPart but unique_ptr is not
-
-        return *(mRootModelPartMap[ModelPartName].get());
+    std::vector< std::string > subparts_list = SplitSubModelPartHierarchy(ModelPartName);
```
Comment (sunethwarna) [inline kratos/containers/model.cpp]:
```suggestion
    const std::vector< std::string >& subparts_list = SplitSubModelPartHierarchy(ModelPartName);
```

File: kratos/containers/model.cpp [hunk 4]
```diff
@@ -33,27 +34,57 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
+        auto search = mRootModelPartMap.find(ModelPartName);
+        if (search == mRootModelPartMap.end()) {
+            CreateRootModelPart(ModelPartName, NewBufferSize);
+            return *(mRootModelPartMap[ModelPartName].get());
+        } else {
+            KRATOS_ERROR << "trying to create a root modelpart with name " << ModelPartName << " however a ModelPart with the same name already exists";
+        }
     } else {
-        KRATOS_ERROR << "trying to create a root modelpart with name " << ModelPartName << " however a ModelPart with the same name already exists";
+        const std::string root_mp_name = subparts_list[0];
```
Comment (sunethwarna) [inline kratos/containers/model.cpp]:
```suggestion
        const std::string& root_mp_name = subparts_list[0];
```

File: kratos/sources/model_part.cpp [hunk 5]
```diff
@@ -1616,28 +1632,71 @@ void ModelPart::RemoveGeometryFromAllLevels(std::string GeometryName)
...
+    } else {
+        ModelPart* p_current_mp = this;
+        for (const auto& r_smp_name : sub_model_parts_list) {
+            if (!p_current_mp->HasSubModelPart(r_smp_name)) {
+                p_current_mp = &p_current_mp->CreateSubModelPart(r_smp_name);
+            } else {
+                p_current_mp = &p_current_mp->GetSubModelPart(r_smp_name);
+            }
+        }
+        return *p_current_mp;
```
Comment (sunethwarna) [inline kratos/sources/model_part.cpp]: One problem which I see in here is, if someone requests to create a submodel part with `NewSubModelPartName = "model_part_1.model_part_2"` where both alreaady exists, then this algorithm will return the existing `model_part_2`, and not throwing an error which is inconsistent with the call `model_part_1.CreateSubModelPart("model_part_2")`. Is there a specific reason why you don't do it in a recursive manner? Then this won't be a problem. Or may be to loop until `N-1` and always use `CreateSubModelPart` on the last.

File: kratos/sources/model_part.cpp [hunk 6]
```diff
@@ -1616,28 +1632,71 @@ void ModelPart::RemoveGeometryFromAllLevels(std::string GeometryName)
...
+                        << "The the following sub model parts are available:";
+                for (const auto& r_avail_smp_name : p_current_mp->GetSubModelPartNames()) {
+                    err_msg << "\n\t" << r_avail_smp_name;
+                }
+                KRATOS_ERROR << err_msg.str() << std::endl;
+            }
+            p_current_mp = &p_current_mp->GetSubModelPart(r_smp_name);
+        }
+        return *p_current_mp;
+    }
```
Comment (sunethwarna) [inline kratos/sources/model_part.cpp]: Or something simpler like
```c++
    const auto delim_pos = SubModelPartName.find('.');
    const std::string& sub_model_part_name = SubModelPartName.substr(0, delim_pos);

    SubModelPartIterator i = mSubModelParts.find(sub_model_part_name);
    KRATOS_ERROR_IF(i == mSubModelParts.end()) << "There is no sub model part with name: \"" << sub_model_part_name << "\" in model part\"" << Name() << "\"" << std::endl;

    if (delim_pos == std::string::npos) {
        return *i;
    } else {
        return i->GetSubModelPart(SubModelPartName.substr(delim_pos + 1));
    }
```

File: kratos/sources/model_part.cpp [hunk 7]
```diff
@@ -1616,28 +1632,71 @@ void ModelPart::RemoveGeometryFromAllLevels(std::string GeometryName)
...
 }
 
 ModelPart* ModelPart::pGetSubModelPart(std::string const& SubModelPartName)
```
Comment (sunethwarna) [inline kratos/sources/model_part.cpp:2011]: I would make this also in the same way as in the reference version

**Review by sunethwarna — COMMENTED** [review]
Thanks @philbucher . Minor comments, and some remarks about the following.

1. `CreateSubModelPart` has an inconsistent behaviour if given with a flat name, and with a name with ".".
2. It would be better if`pGetSubModelPart` with `std::string` input is also changed
3. Wouldn't it better if the `RemoveSubModelPart` with `std::string` also behave in the same manner?

**Comment by sunethwarna** [comment]
This is ready to be reviewed @philbucher @rubenzorrilla @loumalouomega @armingeiser @adityaghantasala @roigcarlo :)

I tagged everyone because I also don't like changing `Model` and `ModelPart` much

File: kratos/containers/model.cpp [hunk 8]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
     KRATOS_TRY
 
-    auto search = mRootModelPartMap.find(ModelPartName);
-    if( search == mRootModelPartMap.end()) {
-        auto pvar_list = Kratos::make_intrusive<VariablesList>();
+    KRATOS_ERROR_IF( ModelPartName.empty() ) << "Please don't use empty names (\"\") when creating a ModelPart" << std::endl;
 
-            ModelPart* pmodel_part = new ModelPart(ModelPartName, NewBufferSize, pvar_list, *this );
-        mRootModelPartMap[ModelPartName] = std::unique_ptr<ModelPart>(pmodel_part); //note that i create it separately since Model is friend of ModelPart but unique_ptr is not
+    const auto delim_pos = ModelPartName.find('.');
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    const auto delim_pos = ModelPartName.find(".");
```

File: kratos/containers/model.cpp [hunk 9]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
 void Model::Reset()
 {
     mRootModelPartMap.clear();
 }
 
+void Model::CreateRootModelPart(const std::string ModelPartName, ModelPart::IndexType NewBufferSize)
+{
+    auto pvar_list = Kratos::make_intrusive<VariablesList>();
+
+    ModelPart* pmodel_part = new ModelPart(ModelPartName, NewBufferSize, pvar_list, *this );
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    ModelPart* p_model_part = new ModelPart(ModelPartName, NewBufferSize, p_var_list, *this );
```

File: kratos/containers/model.cpp [hunk 10]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
+    mRootModelPartMap[ModelPartName] = std::unique_ptr<ModelPart>(pmodel_part); // note that i create it separately since Model is friend of ModelPart but unique_ptr is not
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    mRootModelPartMap[ModelPartName] = std::unique_ptr<ModelPart>(p_model_part); // note that i create it separately since Model is friend of ModelPart but unique_ptr is not
```

File: kratos/containers/model.cpp [hunk 11]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
 } // empty namespace for internal functions
 
 void Model::Reset()
 {
     mRootModelPartMap.clear();
 }
 
+void Model::CreateRootModelPart(const std::string ModelPartName, ModelPart::IndexType NewBufferSize)
+{
+    auto pvar_list = Kratos::make_intrusive<VariablesList>();
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    auto p_var_list = Kratos::make_intrusive<VariablesList>();
```

File: kratos/containers/model.cpp [hunk 12]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
+    const auto delim_pos = ModelPartName.find('.');
+    const std::string& root_model_part_name = ModelPartName.substr(0, delim_pos);
 
-        return *(mRootModelPartMap[ModelPartName].get());
+    if (delim_pos == std::string::npos) {
+        if (mRootModelPartMap.find(root_model_part_name) == mRootModelPartMap.end()) {
+            CreateRootModelPart(root_model_part_name, NewBufferSize);
+            return *(mRootModelPartMap[root_model_part_name].get());
+        } else {
+            KRATOS_ERROR << "trying to create a root modelpart with name " << ModelPartName << " however a ModelPart with the same name already exists";
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
            KRATOS_ERROR << "Trying to create a root modelpart with name " << ModelPartName << " however a ModelPart with the same name already exists";
```

File: kratos/containers/model.cpp [hunk 13]
```diff
@@ -33,27 +34,44 @@ void GetNameWithAscendants(const ModelPart& rModelPart, std::vector<std::string>
...
+        }
     } else {
-        KRATOS_ERROR << "trying to create a root modelpart with name " << ModelPartName << " however a ModelPart with the same name already exists";
+        if (mRootModelPartMap.find(root_model_part_name) == mRootModelPartMap.end()) {
+            CreateRootModelPart(root_model_part_name, NewBufferSize);
+
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
```

File: kratos/containers/model.cpp [hunk 14]
```diff
@@ -63,11 +81,11 @@ void Model::DeleteModelPart( const std::string ModelPartName  )
...
     KRATOS_TRY
 
-    if(this->HasModelPart(ModelPartName))
+    if(this->HasModelPart(ModelPartName)) {
         mRootModelPartMap.erase(ModelPartName); //NOTE: the corresponding variable list should NOT be removed
-    else
-        KRATOS_WARNING("Info") << "attempting to delete inexisting modelpart : " << ModelPartName << std::endl;
-
+    } else {
+        KRATOS_WARNING("Model") << "attempting to delete inexisting modelpart : " << ModelPartName << std::endl;
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
        KRATOS_WARNING("Model") << "Attempting to delete inexisting modelpart : " << ModelPartName << std::endl;
```

File: kratos/containers/model.cpp [hunk 15]
```diff
@@ -99,12 +117,12 @@ ModelPart& Model::GetModelPart(const std::string& rFullModelPartName)
     KRATOS_ERROR_IF( rFullModelPartName.empty() ) << "Attempting to find a "
         << "ModelPart with empty name (\"\")!" << std::endl;
 
-    std::vector< std::string > subparts_list = SplitSubModelPartHierarchy(rFullModelPartName);
-
+    const auto delim_pos = rFullModelPartName.find('.');
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    const auto delim_pos = rFullModelPartName.find(".");
```

File: kratos/containers/model.cpp [hunk 16]
```diff
@@ -245,23 +239,19 @@ bool Model::HasModelPart(const std::string& rFullModelPartName) const
     KRATOS_ERROR_IF( rFullModelPartName.empty() ) << "Attempting to find a "
         << "ModelPart with empty name (\"\")!" << std::endl;
 
-    std::vector< std::string > subparts_list = SplitSubModelPartHierarchy(rFullModelPartName);
+    const auto delim_pos = rFullModelPartName.find('.');
```
Comment (loumalouomega) [inline kratos/containers/model.cpp]:
```suggestion
    const auto delim_pos = rFullModelPartName.find(".");
```

File: kratos/containers/model.cpp [hunk 17]
```diff
@@ -117,7 +116,7 @@ ModelPart& Model::GetModelPart(const std::string& rFullModelPartName)
     KRATOS_ERROR_IF( rFullModelPartName.empty() ) << "Attempting to find a "
         << "ModelPart with empty name (\"\")!" << std::endl;
 
-    const auto delim_pos = rFullModelPartName.find('.');
+    const auto delim_pos = rFullModelPartName.find(".");
     const std::string& root_model_part_name = rFullModelPartName.substr(0, delim_pos);
```
Comment (philbucher) [inline kratos/containers/model.cpp]: @loumalouomega why did you change this?

Comment (loumalouomega) [reply kratos/containers/model.cpp]: We use "" for the strings instead of ''

Comment (philbucher) [reply kratos/containers/model.cpp]: aren't those two different things?
'' is a single char and "" is a string no?
@roigcarlo do you know better? :D

Comment (loumalouomega) [reply kratos/containers/model.cpp]: I think, those are the same, but we use " for strings

Comment (sunethwarna) [reply kratos/containers/model.cpp]: AFAIK, `'*'` is a char and `"*"` is a string, but i have no idea whether they have any overhead :D

Comment (loumalouomega) [reply kratos/containers/model.cpp]: Okay, I assume the method find will use a string anyway, and basically does  a cast, so no idea

**Review by rubenzorrilla — APPROVED** [review]
